### DISCUSSION TITLE
How to replace the name `Result`?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#705] [Polls]
Working on the new Rust wrapper forced me to find a new name for `Result` because that is functionally a keyword in Rust. I came up with `ReaderResult` (or maybe actually @SteveCookTU did in his first proposal) which felt most in line with `ReaderOptions`. I am intending to replace the term `Result` across all the APIs mainly for two reasons:

1. `Result` is in fact a bad name because it is way too generic and conveys no meaning.
2. It is too generic even for this trivial API because what is going to be the return type of `WriteBarcode`?

The second reason would lend itself to introduce the terms `ReaderResult` and `WriterResult` but it still fails to fix the first issue. The `MultiFormatWriter.encode()` function returns a `BitMatrix`. That is not going to be the return type of `WriteBarcode` because I want it to support the transfer of some metadata. In that context I thought about naming it `Symbol` because it represents that thing that you would print on paper and most specifications use the term "symbol" for that, so does the `zint` library. Naming one thing `Result` and the other `Symbol` has obviously no consistency what-so-ever.

I have not thought this through fully but I think the best way forward is to have both `ReadBarcode` and `WriteBarcode` return the same type (basically current `Result`) and have that type contain a `symbol` property that allows you get access to the bitmap data you can use for printing it. This could even be of interest for the reading part, if someone wanted to have access to that information which is obviously always generated inside the library during reading/scanning anyway.

Any renaming will only take full effect for the c++ API with the next major version for ABI compatibility reasons (see `DecodeHints` -> `ReaderOptions`). But since neither the C, Rust nor .NET wrappers are released, yet, I'd like to use the future proof name there already.

The obvious replacement for `Result` in my opinion is `Barcode`.

**Comment by markusfisch** [comment]
Hm, just off the top of my head - I wouldn't call the result of reading a barcode a barcode, but content or data or information.

We give the reader an image of a barcode and get the decoded content(s?).

Conversely, we give the writer content or data to encode it in a barcode.

**Reply by axxel** [reply]
But `ReadBarcodes` does not take 'a barcode' but an image as input and it tries to find all the barcode (symbols) there are in that image and then returns something that provides access to a bunch of properties, like the 'content' (in form of `text`, `textECI`, `bytes`, `bytesECI`), 'format', 'position'. And those properties are the properties of 'a barcode' and not of 'some data'.

Maybe it would feel more natural to you if you imagined the function being called `DetectBarcodes`? I decided against that a couple years ago since the verb 'to detect' did already have some specific meaning inside the code base and also because I liked the symmetry of *reading* and *writing* barcodes, even though something like *scanning* and *generating* barcodes sounded a little more fitting for the domain.

**Reply by axxel** [reply]
And also: the `ReadBarcode` function was replacing the `MultiFormatReader` API... (to put credit where it's due ;)).

**Reply by markusfisch** [reply]
I see what you mean, and from that perspective `Barcode` is probably a good umbrella term, especially when you're planning to use the same type for the reader and writer for symmetry.

Still, I would prefer something like `BarcodeData`, which would make its nature more clear, and to me would be a more intuitive result of `ReadBarcode`.

**Comment by Sec-ant** [comment]
Just to make sure I didn't misunderstand it, if I vote for `Barcode`, I'm voting for it to be used as the name of the type of those `read`/`write` functions outputs. And it'll be a type that will contain some metadata of reading/writing, possibly including a `symbol` property for users to access the bitmap data of the symbol that is read or written?

**Reply by axxel** [reply]
That is exactly my current idea/concept. But note the "have not fully thought this through" part ;).

**Reply by Sec-ant** [reply]
Then I like the `Barcode` name and I kinda like the whole idea. FWIW, there's a similar term in the [draft spec of the Barcode Detection API in web platforms](https://wicg.github.io/shape-detection-api/#barcode-detection-api) that uses the name [`DetectedBarcode`](https://wicg.github.io/shape-detection-api/#dictdef-detectedbarcode) for the output of a barcode detection. So considering we're to use it for both read and write functions, `Barcode` seems a natural choice to me. `Result` is too general and I think `Symbol` is a very strange word to most of the end users, in addition that we may plan to use `symbol` as a property to refer to the bitmap image.

And I also like the counterpart relationship of **read** and **write**.

One thing I'm not certain of though is that is there the necessity or a practical reason behind that we use the same type for the outputs of both read and write functions?

**Reply by axxel** [reply]
> DetectedBarcode

We would have to use the term "ReadBarcode" then, as in "simple passed tense of 'to read'", which due to the funny coincidence of having both spelled exactly the same in English, is a non-starter... But thanks for the hint to that spec.

> One thing I'm not certain of though is that is there the necessity or a practical reason behind that we use the same type for the outputs of both read and write functions?

That idea was born from the realization that `BitMatrix` is too basic as a proper return type of `WriteBarcode`. And since it is pretty conceivable why someone would want to e.g. check the `ecLevel` of a newly written barcode or someone else would like to get access to the sampled BitMatrix (the symbol) of a read barcode, then it makes totally sense to use the same type. There are a few properties that don't make much sense in the one or the other case (like position) but I'm convinced that the increased simplicity by using the same type far outweighs that.

**Comment by siiky** [comment]
To me, `Barcode` doesn't convey the fact that `ReadBarcode` can fail, so I voted for `ReadResult` which seems to be the most explicit. Additionally it's similar to the current name so it should be easy(ier?) for programmers to pick the change.

But whatever option you choose is fine with me.

**Reply by axxel** [reply]
Well, especially on the C-API the returned value of `zxing_ReadBarcode` is a pointer and if that fails, you get `NULL` (see also my other comment about the possibility to change that...). And the interpretation of `NULL` is rather universal regardless of what the type the pointer is usually pointing to, right?

Anyway: I did not get any feedback that convinced me that there is a better name than `Barcode`. So the C/Rust/.NET wrappers use this term now, the rest will follow later. Thanks for your input, nonetheless.

**Comment by benjohnde** [comment]
I highly believe names should reflect both the output and how it was derived properly. I am very strong for `Barcode`. I don't want to open documentation if I can avoid it. Having a fluent interface is beneficial in achieving that.

**Comment by axxel** [comment]
Update: I've done the `Result` -> `Barcode` rename throughout the c++ source code [here](https://github.com/zxing-cpp/zxing-cpp/commit/bdf97f4c0717cd40ba063652a87b54ee5247ba67). The class itself it still called `Result` to keep the ABI stable (not making the same mistake as I did with the `DecodeHints` -> `ReaderOptions` renaming). With the next major release, the class will be renamed.

**Reply by axxel** [reply]
This rename now happened and will be part of the 3.0 release.
